tensorrt_llm/_torch/models/modeling_speculative.py

```
@@ -989,6 +1139,34 @@ def project_target_hidden(self,
         hidden_states = hidden_states.to(self.fc.weight.dtype)
         return self.hidden_norm(self.fc(hidden_states))
 
+    @property
+    def has_markov_head(self) -> bool:
+        return self._dspark_markov_rank > 0 and self.markov_w1 is not None
+
+    def apply_markov_chain_logits(self,
+                                  base_logits: torch.Tensor,
+                                  first_prev_tokens: torch.Tensor,
+                                  argmax_fn=None,
+                                  vocab_slice: slice = None) -> torch.Tensor:
```

Ruff (0.16.1)

[warning] PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)